### PR TITLE
Game Lab: Create new GameLabP5 object, allow p5 to block on draw() and setup()

### PR DESCRIPTION
**Pull request by cpirich**
- Refactor all of the the wrapping and patching of p5/p5play into a new `GameLabP5` object
- Modified the way we handle `setup()` and `draw()` from p5 so that these can be long running calls (either due to expensive user code or because the code inside those methods is stopped in the debugger).
  1. Accomplished this by patching p5's `redraw()` and `_setup()` methods to make them essentially asynchronous. New methods were created on p5 containing the code that runs after `userDraw()` and `userSetup()` are called, respectively.
  2. Our Game Lab code has been modified to track when a `draw()` or `setup()` call has started (using `this.drawInProgress` and `this.setupInProgress`). The initial attempts to call those methods check to see if they need to call the `afterDrawComplete()` or `afterSetupComplete()` methods afterwards. Since the user code may not be complete, we also check for draw/setup completion after we call `executeInterpreter()` in the `onTick()` loop.

### CODE REVIEW

File: apps/src/gamelab/GameLab.js [hunk 1]
```diff
@@ -42,15 +43,10 @@ var GameLab = function () {
...
-  this.p5decrementPreload = null;
-  this.p5eventNames = [
-    'mouseMoved', 'mouseDragged', 'mousePressed', 'mouseReleased',
-    'mouseClicked', 'mouseWheel',
-    'keyPressed', 'keyReleased', 'keyTyped'
-  ];
-  this.p5specialFunctions = ['preload', 'draw', 'setup'].concat(this.p5eventNames);
-
+  this.drawInProgress = false;
+  this.setupInProgress = false;
```
Comment (islemaster) [inline apps/src/gamelab/GameLab.js:47]: Just a thought: It seems unlikely to happen, but would it be harmful for both of these to be `true` at the same time?  If so, would it be better to model them as a single state?

File: apps/src/gamelab/GameLabP5.js [hunk 2]
```diff
@@ -0,0 +1,435 @@
+'use strict';
```
Comment (islemaster) [inline apps/src/gamelab/GameLabP5.js:1]: Shameless plug: I like to add [these stricter linting rules](https://github.com/code-dot-org/code-dot-org/blob/staging/apps/src/applab/EventSandboxer.js#L2) to new files, as we gradually work toward making them global linting rules. :grin:

File: apps/src/gamelab/GameLabP5.js [hunk 3]
```diff
@@ -0,0 +1,435 @@
...
+var assetPrefix = require('../assetManagement/assetPrefix');
+
+/**
+ * An instantiable GameLabP5 class that wraps p5 and p5play and patches it in
+ * specific places to enable GameLab functionality
+ */
+var GameLabP5 = function () {
```
Comment (islemaster) [inline apps/src/gamelab/GameLabP5.js:8]: Yay for clean constructors!

**Comment by islemaster** [comment]
This looks _fantastic!_  Thanks for writing this wrapper.  It's really nice to have some of the weirder stuff we're doing with p5 hidden away in here, and move more code out of gamelab.js.

LGTM!

File: apps/src/gamelab/GameLab.js (hunk 1 above)
Comment (cpirich) [inline apps/src/gamelab/GameLab.js:47]: That's a good idea. These calls should not overlap. Will do that in a future PR.

**Pull request merged**